ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedModule.java

```diff
@@ -0,0 +1,228 @@
+  private ArrayList<UIThreadOperation> mOperations = new ArrayList<>();
+  private volatile @Nullable ArrayList<UIThreadOperation> mReadyOperations = null;
+
+  /*
+   * Nodes manager will only be accessed from the UI Thread
+   */
+  private final NativeAnimatedNodesManager mNodesManager = new NativeAnimatedNodesManager();
+
+  public NativeAnimatedModule(ReactApplicationContext reactContext) {
+    super(reactContext);
+  }
+
+  @Override
+  public void initialize() {
+    // Safe to acquire choreographer here, as initialize() is invoked from UI thread.
+    mReactChoreographer = ReactChoreographer.getInstance();
+
+    ReactApplicationContext reactCtx = getReactApplicationContext();
+    final NativeViewHierarchyManager nativeViewHierarchyManager =
+      reactCtx.getNativeModule(UIManagerModule.class).getNativeViewHierarchyManager();
+    mAnimatedFrameCallback = new GuardedChoreographerFrameCallback(reactCtx) {
+      @Override
+      protected void doFrameGuarded(final long frameTimeNanos) {
+
+        ArrayList<UIThreadOperation> operations;
+        synchronized (mOperationsCopyLock) {
+          operations = mReadyOperations;
+          mReadyOperations = null;
+        }
+
+        if (operations != null) {
+          for (int i = 0, size = operations.size(); i < size; i++) {
+            operations.get(i).execute(mNodesManager);
+          }
+        }
+
+        if (mNodesManager.hasActiveAnimations()) {
+          mNodesManager.runUpdates(nativeViewHierarchyManager, frameTimeNanos);
+        }
+
+        Assertions.assertNotNull(mReactChoreographer).postFrameCallback(
+          ReactChoreographer.CallbackType.ANIMATIONS,
+          mAnimatedFrameCallback);
+      }
+    };
+    reactCtx.addLifecycleEventListener(this);
+  }
+
+  @Override
+  public void onBatchComplete() {
+    // Note: The order of executing onBatchComplete handler (especially in terms of onBatchComplete
+    // from the UIManagerModule) doesn't matter as we only enqueue operations for the UI thread to
+    // be executed from here. Thanks to ReactChoreographer all the operations from here are going
+    // to be executed *after* all the operations enqueued by UIManager as the callback type that we
+    // use for ReactChoreographer (CallbackType.ANIMATIONS) is run after callbacks that UIManager
+    // use
+    ArrayList<UIThreadOperation> operations = mOperations.isEmpty() ? null : mOperations;
+    if (operations != null) {
+      mOperations = new ArrayList<>();
+      synchronized (mOperationsCopyLock) {
+        if (mReadyOperations == null) {
+          mReadyOperations = operations;
+        } else {
+          mReadyOperations.addAll(operations);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void onHostResume() {
+    enqueueFrameCallback();
+  }
+
+  @Override
+  public void onHostPause() {
+    clearFrameCallback();
+  }
+
+  @Override
+  public void onHostDestroy() {
+    clearFrameCallback();
```

Comment:
I think we should just remove it from onHostDestroy then -- it's confusing otherwise and we shouldn't really be trying to handle cases that we don't think are correct to happen in the first place